dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/TaskSetState.java

```diff
@@ -16,60 +16,135 @@
  */
 package com.google.edwmigration.dumper.application.dumper.task;
 
+import static com.google.edwmigration.dumper.application.dumper.task.TaskCategory.REQUIRED;
+import static com.google.edwmigration.dumper.application.dumper.task.TaskState.FAILED;
+
+import com.google.common.collect.ImmutableList;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.stream.Collectors;
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
 import javax.annotation.concurrent.GuardedBy;
 import net.jcip.annotations.ThreadSafe;
 
 /** @author shevek */
 public interface TaskSetState {
```

Comment:
@paolomorandini Let's continue the discussion here, because github won't allow in-thread replies unless a comment is attached to a line of code.

- I've reinstated `@GuardedBy` with `this` as suggested. The annotation doesn't seem to work in my setup, but if it works for some people I don't mind keeping it.
- I don't have an opinion on this pattern in general, I agree that in this case it's not necessary
- There is room for improvement of the performance of this class, including less work within `synchronized` blocks, but I think it would be better to do this separately from the current PR and only if we have flows with a sufficiently large number of Tasks.

Returning a copy of the map would work, but I don't want to make the map part of the api, because it will prevent future optimizations. For example, `failedRequiredTaskCount` could be optimized to constant-time if we track it separately, but returning a copy of the whole map will always be linear.